crates/lib/src/bootc_composefs/update.rs

```diff
@@ -1,33 +1,153 @@
     cli::UpgradeOpts,
+    spec::ImageReference,
+    store::ComposefsRepository,
 };
 
+use cap_std_ext::cap_std::{ambient_authority, fs::Dir};
+
+#[context("Getting SHA256 Digest for {id}")]
+pub fn str_to_sha256digest(id: &str) -> Result<Sha256Digest> {
+    let id = id.strip_prefix("sha256:").unwrap_or(id);
+    Ok(parse_sha256(&id)?)
+}
+
+/// Checks if a container image has been pulled to the local composefs repository.
+///
+/// This function verifies whether the specified container image exists in the local
+/// composefs repository by checking if the image's configuration digest stream is
+/// available. It retrieves the image manifest and configuration from the container
+/// registry and uses the configuration digest to perform the local availability check.
+///
+/// # Arguments
+///
+/// * `repo` - The composefs repository
+/// * `imgref` - Reference to the container image to check
+///
+/// # Returns
+///
+/// Returns a tuple containing:
+/// * `true` if the image is pulled/available locally, `false` otherwise
+/// * The container image manifest
+/// * The container image configuration
+#[context("Checking if image {} is pulled", imgref.image)]
+async fn is_image_pulled(
+    repo: &ComposefsRepository,
+    imgref: &ImageReference,
+) -> Result<(bool, ImageManifest, ImageConfiguration)> {
+    let imgref_repr = get_imgref(&imgref.transport, &imgref.image);
+    let (manifest, config) = get_container_manifest_and_config(&imgref_repr).await?;
+
+    let img_digest = manifest.config().digest().digest();
+    let img_sha256 = str_to_sha256digest(&img_digest)?;
+
+    // check_stream is expensive to run, but probably a good idea
+    let container_pulled = repo.check_stream(&img_sha256).context("Checking stream")?;
+
+    Ok((container_pulled.is_some(), manifest, config))
+}
+
 #[context("Upgrading composefs")]
-pub(crate) async fn upgrade_composefs(_opts: UpgradeOpts) -> Result<()> {
+pub(crate) async fn upgrade_composefs(opts: UpgradeOpts) -> Result<()> {
     let host = composefs_deployment_status()
         .await
         .context("Getting composefs deployment status")?;
 
-    start_finalize_stated_svc()?;
-
-    // TODO: IMPORTANT We need to check if any deployment is staged and get the image from that
-    let imgref = host
+    let mut imgref = host
         .spec
         .image
         .as_ref()
         .ok_or_else(|| anyhow::anyhow!("No image source specified"))?;
 
+    let sysroot =
+        Dir::open_ambient_dir("/sysroot", ambient_authority()).context("Opening sysroot")?;
+    let repo = open_composefs_repo(&sysroot)?;
+
+    let (img_pulled, mut manifest, mut config) = is_image_pulled(&repo, imgref).await?;
+    let booted_img_digest = manifest.config().digest().digest();
+
+    // We already have this container config. No update available
+    if img_pulled {
+        println!("No changes in: {imgref:#}");
+        // TODO(Johan-Liebert1): What if we have the config but we failed the previous update in the middle?
```

Comment:
Yes, I think we need to only write the config object after all of its dependent layers.